File: src/main/java/at/searles/regexp/CharSet.java
package at.searles.regexp;

import at.searles.lexer.utils.Interval;
import at.searles.lexer.utils.IntervalSet;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.Iterator;

public class CharSet implements Regexp, Iterable<Interval> {

    private final IntervalSet<Void> set;

    private CharSet(IntervalSet<Void> set) {
        this.set = set;
    }

    private CharSet() {
        this(new IntervalSet<>());
    }

    public static CharSet chars(int... chars) {
        Arrays.sort(chars);

        IntervalSet<Void> set = new IntervalSet<>();

        for (int ch : chars) {
            set.add(ch, ch + 1, null, (a, b) -> null);
        }

        return new CharSet(set);
    }

    /**
     * @param intervals [a, b], b is inclusive!
     * @return The CharSet convaining the intervals provided
     */
    public static CharSet interval(int... intervals) {
        if (intervals.length % 2 != 0) {
            throw new IllegalArgumentException("must have an even number of ranges");
        }

        IntervalSet<Void> set = new IntervalSet<>();

        for (int i = 0; i < intervals.length; i += 2) {
            set.add(intervals[i], intervals[i + 1] + 1, null, (a, b) -> null);
        }

        return new CharSet(set);
    }

    public static CharSet empty() {
        return new CharSet();
    }

    /**
     * Recognizes all positive integers except for Integer.MAX_VALUE.
     */
    public static CharSet all() {
        return interval(0, Integer.MAX_VALUE - 1);
    }

    public <A> IntervalSet<A> copyIntervalSet(A a) {
        return set.copy(ignore -> a);
    }

    public boolean isAll() {
        IntervalSet.Iter<Void> it = set.iterator();

        if (it.hasNext()) {
            it.next();

            return it.start() == 0 && it.end() == Integer.MAX_VALUE;
        }

        return false;
    }

    public boolean isEmpty() {
        return set.isEmpty();
    }

    public CharSet union(CharSet that) {
        return new CharSet(new IntervalSet<>(set).add(that.set, (a, b) -> null));
    }

    public boolean isInverted() {
        if (!set.isEmpty()) {
            IntervalSet.Iter<Void> it = set.iterator();
            it.previous();

            return it.end() == Integer.MAX_VALUE;
        }

        return false;
    }

    public CharSet invert() {
        IntervalSet<Void> inverted = new IntervalSet<>();

        int start = 0;

        IntervalSet.Iter<Void> it = set.iterator();

        while (it.hasNext()) {
            it.next();
            inverted.add(start, it.start(), null, (a, b) -> null);
            start = it.end();
        }

        inverted.add(start, Integer.MAX_VALUE, null, (a, b) -> null);
        return new CharSet(inverted);
    }

    @Override
    public <A> A accept(Visitor<A> visitor) {
        return visitor.visitCharSet(this);
    }

    @Override
    public String toString() {
        return String.format("CharSet(%s)", set.toString());
    }

    public boolean contains(int ch) {
        return set.contains(ch);
    }

    @NotNull
    @Override
    public Iterator<Interval> iterator() {
        return new Iterator<Interval>() {

            final IntervalSet.Iter<Void> it = set.iterator();

            @Override
            public boolean hasNext() {
                return it.hasNext();
            }

            @Override
            public Interval next() {
                it.next();
                return new Interval(it.start(), it.end());
            }
        };
    }
}

File: src/main/java/at/searles/regexp/CharSet.kt
package at.searles.regexp

import at.searles.lexer.utils.Interval
import at.searles.lexer.utils.IntervalSet
import java.util.*

class CharSet private constructor(private val set: IntervalSet) : Regexp, Iterable<Interval> {
    fun union(that: CharSet): CharSet {
        return CharSet(set.copy().apply { add(that.set) })
    }

    fun invert(): CharSet {
        return CharSet(set.inverted(0, Int.MAX_VALUE))
    }

    override fun <A> accept(visitor: Visitor<A>): A {
        return visitor.visitSet(set)
    }

    override fun toString(): String {
        return String.format("CharSet(%s)", set.toString())
    }

    operator fun contains(ch: Int): Boolean {
        return set.contains(ch)
    }

    override fun iterator(): Iterator<Interval> {
        return set.iterator()
    }

    companion object {
        fun chars(vararg chars: Char): CharSet {
            Arrays.sort(chars)
            val set = IntervalSet()

            chars.forEach {
                set.add(Interval(it.toInt(), it.toInt() + 1))
            }

            return CharSet(set)
        }

        fun chars(vararg ints: Int): CharSet {
            Arrays.sort(ints)
            val set = IntervalSet()

            ints.forEach {
                set.add(Interval(it, it + 1))
            }

            return CharSet(set)
        }

        /**
         * @param intervals [a, b], b is inclusive!
         * @return The CharSet convaining the intervals provided
         */
        fun interval(vararg intervals: Int): CharSet {
            require(intervals.size % 2 == 0) { "must have an even number of ranges" }
            val set = IntervalSet()

            var i = 0

            while (i < intervals.size) {
                set.add(Interval(intervals[i], intervals[i + 1] + 1))
                i += 2
            }

            return CharSet(set)
        }

        fun empty(): CharSet {
            return CharSet(IntervalSet())
        }

        fun all(): CharSet {
            return CharSet(IntervalSet().apply { add(Interval(0, Int.MAX_VALUE)) })
        }

        fun eof(): Regexp {
            return chars(-1)
        }
    }
}